### DISCUSSION TITLE
readonly local variable

### DISCUSSION THREAD

**Discussion by HakamFostok** [dotnet/csharplang#5057] [Language Ideas] [answered]
# Readonly local variable

## Summary
Allow local variables in the methods to be set only once.

## Motivation
Some local variables in the method (maybe most of them) are set only once. 
After they had been set their value does not change, and maybe changing the value after it has been set could cause a problem in the logic of the program.
It could be beneficial to make this explicitly by specifying the variable as `readonly`.
Right now, this is supported for the parameters of the method by specifying the `in` keyword for the parameter.
And also local variables can have the `const` modifier to achieve what is proposed here, but the problem with the `const ` keyword on the local variable is that the value to set must be compile-time constant.
Adding the ability to mark local variables as `readonly` will complete the circuit of the `in` keywords for the parameter and the `const` keywords for the local variables.

This will make the intention of the code clearer when it is read by another person.

## Detailed design

The design is very simple here is an example

```
public int SomeMethod()
{
     readonly int result = 1;    // After this point the result variable can not be set
     result = 2;                        // Will cause a compile-time error
     return result;
}
```

## Drawbacks
Maybe some people may say that this adding a feature to C# and make it more complex language.
In most general cases, and personally, I do not buy into this discussion, I think the language must keep evolving and adding new features, and it is the responsibility of the developer to keep learning.
Plus I think this is a very simple feature and not complex at all to learn.
Above all that it is optional, it can be used, only if the developer wants to be more defensive and clear with the purpose of the code.

## Alternatives
Maybe expanding the scope of the already-exists functionality of the `const` keyword on the local variables could be possible.
So the `const` local variable will allow the value to be a none-compile-time constant.
But I do not know if this could be considered as a breaking change or not.
Actually, the old programs will not break so I think, just in my opinion, it is not a breaking change.

## Unresolved questions

**Comment by khm1600** [comment, marked as answer]
https://github.com/dotnet/csharplang/issues/188

**Reply by HakamFostok** [reply]
Thank you @khm1600 
I this coming in C# 10 ?

**Reply by HaloFour** [reply]
This feature has not been slated for any particular release of C#, if it is implemented at all.  Being "championed" means that there are language team members who are willing to pursue a design and to bring it to language design meetings.

**Reply by svick** [reply]
@HakamFostok No, it's not. That feature is currently on the backlog, so it might come in some future release after C# 10.
